File: python/src/wh40kdc/imports/gw.py
"""GW adapter: lower the Games Workshop 40K app's plain-text army-list export
to a ``ParsedRoster``.

The format opens with the same ``++++…++++`` summary fence as the NewRecruit
WTC formats, then lists units grouped under ALL-CAPS battlefield-role
sections. Each unit is a header line ``Name (N pts)`` followed by
``•``-bulleted entries.

Bullet classification (the parsing crux):

- A top-level ``• Nx Thing`` *with* further-indented child bullets is a
  **model group** — ``N`` adds to the model count and the children are that
  group's wargear.
- A top-level ``• Nx Thing`` *without* children is plain **wargear**.
- A bullet *without* an ``Nx`` count is an **annotation**: ``… Character``
  flags a character, ``Warlord`` flags the warlord, ``Name (+N pts)`` is the
  enhancement.

**Disjointness from the WTC matchers**: the GW format always carries ``•``
bullets and never the WTC ``N with`` lines.

Python mirror of ``tools/src/import/gw.ts``.
"""

from __future__ import annotations

import re
from typing import Any

from wh40kdc.imports.adapter import FormatAdapter
from wh40kdc.imports.newrecruit_text import (
    faction_from_keyword,
    infer_battle_size_raw,
    strip_parenthetical,
)

_FACTION_KEYWORD_PREFIX = "+ FACTION KEYWORD:"

_HEADER_FACTION = re.compile(r"^\+\s*FACTION KEYWORD:\s*(.+?)\s*$", re.IGNORECASE)
_HEADER_DETACHMENT = re.compile(r"^\+\s*DETACHMENT:\s*(.+?)\s*$", re.IGNORECASE)
_HEADER_TOTAL_POINTS = re.compile(r"^\+\s*TOTAL ARMY POINTS:\s*(\d+)\s*pts?\s*$", re.IGNORECASE)

_FENCE = re.compile(r"^\++\s*$")
_HEADER_LINE = re.compile(r"^\+")
_SECTION_HEADER = re.compile(r"^[A-Z][A-Z0-9 \-/&]+$")  # BATTLELINE, ALLIED UNITS, …
_UNIT_HEADER = re.compile(r"^(.+?)\s*\(\s*(\d+)\s*pts?\s*\)\s*$", re.IGNORECASE)
_BULLET_LINE = re.compile(r"^(\s*)•\s*(.+?)\s*$")
_NX_PREFIX = re.compile(r"^(\d+)x\s+(.+)$")
_ENHANCEMENT_ANNOT = re.compile(r"^(.+?)\s*\(\+\s*(\d+)\s*pts?\s*\)\s*$", re.IGNORECASE)
_WITH_LINE = re.compile(r"^[\t ]*\d+\s+with\b", re.MULTILINE)
_BULLET = re.compile(r"^[\t ]*•", re.MULTILINE)
_SPLIT_LINES = re.compile(r"\r?\n")

_ALLIED_SECTION = "ALLIED UNITS"
_CHARACTERS_SECTION = "CHARACTERS"
_CHARACTER_SUFFIX = " Character"
_WARLORD_MARKER = "Warlord"


def _is_gw_text(decoded: Any) -> str | None:
    """Accept the input only when it carries the FACTION KEYWORD summary
    header, has ``•`` bullets, and lacks the WTC ``N with`` body lines."""
    if not isinstance(decoded, str):
        return None
    if _FACTION_KEYWORD_PREFIX not in decoded:
        return None
    if _BULLET.search(decoded) is None:
        return None
    if _WITH_LINE.search(decoded) is not None:  # that's wtc-full
        return None
    return decoded


def _parse_header(lines: list[str]) -> tuple[dict[str, Any], int] | None:
    faction_raw_name: str | None = None
    detachment_raw_name: str | None = None
    total_reported: int | None = None

    fence_indices: list[int] = []
    for i, line in enumerate(lines):
        if len(fence_indices) >= 2:
            break
        if _FENCE.match(line):
            fence_indices.append(i)

    saw_faction_keyword = False
    for line in lines:
        if not line.startswith("+"):
            continue
        m = _HEADER_FACTION.match(line)
        if m:
            faction_raw_name = faction_from_keyword(m.group(1))
            saw_faction_keyword = True
            continue
        m = _HEADER_DETACHMENT.match(line)
        if m:
            detachment_raw_name = strip_parenthetical(m.group(1))
            continue
        m = _HEADER_TOTAL_POINTS.match(line)
        if m:
            total_reported = int(m.group(1))

    if not saw_faction_keyword:
        return None

    body_start = fence_indices[1] + 1 if len(fence_indices) >= 2 else 0
    # The GW export has no POINTS LIMIT line — only TOTAL ARMY POINTS. Use it
    # as the declared limit so the inferred battle size stays round-trippable.
    declared_limit = total_reported
    header = {
        "name": "Imported roster",
        "faction_raw_name": faction_raw_name,
        "detachment_raw_name": detachment_raw_name,
        "total_reported": total_reported,
        "declared_limit": declared_limit,
        "battle_size_raw": infer_battle_size_raw(declared_limit),
    }
    return header, body_start


def _finish_unit(acc: dict[str, Any]) -> dict[str, Any]:
    bullets: list[dict[str, Any]] = acc["bullets"]
    top_indent = min((b["indent"] for b in bullets), default=0)

    wargear: dict[str, int] = {}
    model_count = 0
    is_warlord = False
    is_character = acc["section"] == _CHARACTERS_SECTION
    enhancement_raw_name: str | None = None
    enhancement_points: int | None = None

    for i, b in enumerate(bullets):
        # A child bullet (deeper than the unit's top level) is a model group's
        # weapon — its `Nx` count is already the squad-wide total.
        if b["indent"] > top_indent:
            if b["count"] is not None:
                wargear[b["text"]] = wargear.get(b["text"], 0) + b["count"]
            continue

        # Top-level annotation (no `Nx` count): enhancement / character / warlord.
        if b["count"] is None:
            enh = _ENHANCEMENT_ANNOT.match(b["text"])
            if enh:
                if enhancement_raw_name is None:
                    enhancement_raw_name = enh.group(1).strip()
                    enhancement_points = int(enh.group(2))
                continue
            for token in (s.strip() for s in b["text"].split(",")):
                if not token:
                    continue
                if token == _WARLORD_MARKER:
                    is_warlord = True
                elif token.endswith(_CHARACTER_SUFFIX):
                    is_character = True
            continue

        # Top-level `Nx` bullet: a model group when it has child bullets
        # beneath it, otherwise plain wargear.
        next_bullet = bullets[i + 1] if i + 1 < len(bullets) else None
        if next_bullet is not None and next_bullet["indent"] > top_indent:
            model_count += b["count"]
        else:
            wargear[b["text"]] = wargear.get(b["text"], 0) + b["count"]

    if model_count == 0:
        model_count = 1

    # The GW unit header points include the enhancement; back it out to the base.
    displayed = acc["displayed_pts"]
    if displayed is None:
        points = None
    elif enhancement_points is not None:
        points = displayed - enhancement_points
    else:
        points = displayed

    return {
        "raw_name": acc["raw_name"],
        "is_character": is_character,
        "model_count": model_count,
        "points": points,
        "is_warlord": is_warlord,
        "enhancement_raw_name": enhancement_raw_name,
        "enhancement_points": enhancement_points,
        "wargear": [{"raw_name": n, "count": c} for n, c in wargear.items()],
    }


def _parse_body(lines: list[str], body_start: int) -> tuple[list[dict[str, Any]], bool]:
    units: list[dict[str, Any]] = []
    current: dict[str, Any] | None = None
    section: str | None = None
    allied_units = 0

    def finalize() -> None:
        nonlocal current
        if current is not None:
            units.append(_finish_unit(current))
            current = None

    for raw in lines[body_start:]:
        line = raw.strip()
        if not line or _FENCE.match(line) or _HEADER_LINE.match(line):
            continue

        bullet_match = _BULLET_LINE.match(raw)
        if bullet_match:
            if current is not None:
                indent = len(bullet_match.group(1))
                rest = bullet_match.group(2)
                nx = _NX_PREFIX.match(rest)
                current["bullets"].append(
                    {
                        "indent": indent,
                        "count": int(nx.group(1)) if nx else None,
                        "text": (nx.group(2) if nx else rest).strip(),
                    }
                )
            continue

        unit_match = _UNIT_HEADER.match(line)
        if unit_match:
            finalize()
            current = {
                "raw_name": unit_match.group(1).strip(),
                "displayed_pts": int(unit_match.group(2)),
                "section": section,
                "bullets": [],
            }
            if section == _ALLIED_SECTION:
                allied_units += 1
            continue

        if _SECTION_HEADER.match(line):
            finalize()
            section = line

    finalize()
    return units, allied_units > 0


def _matches(decoded: Any) -> bool:
    return _is_gw_text(decoded) is not None


def _parse(decoded: Any) -> dict[str, Any]:
    text = _is_gw_text(decoded)
    if text is None:
        raise ValueError("gw: input is not a GW app text export")

    lines = _SPLIT_LINES.split(text)
    parsed = _parse_header(lines)
    if parsed is None:
        raise ValueError('gw: missing "+ FACTION KEYWORD:" header')
    header, body_start = parsed

    units, multi_force = _parse_body(lines, body_start)

    total_computed = 0
    for u in units:
        total_computed += u["points"] or 0
        total_computed += u["enhancement_points"] or 0

    return {
        "name": header["name"],
        "generated_by": None,
        "faction_raw_name": header["faction_raw_name"],
        "detachment_raw_name": header["detachment_raw_name"],
        "battle_size_raw": header["battle_size_raw"],
        "declared_limit": header["declared_limit"],
        "total_reported": header["total_reported"],
        "total_computed": total_computed,
        "units": units,
        "multi_force": multi_force,
    }


gw_adapter = FormatAdapter(id="gw", matches=_matches, parse=_parse)

File: tools/src/import/gw.ts
/**
 * GW adapter: lower the Games Workshop 40K app's plain-text army-list export to
 * a {@link ParsedRoster}.
 *
 * The format opens with the same `++++…++++` summary fence as the NewRecruit WTC
 * formats (FACTION KEYWORD / DETACHMENT / TOTAL ARMY POINTS / WARLORD /
 * ENHANCEMENT / NUMBER OF UNITS / SECONDARY), then lists units grouped under
 * ALL-CAPS battlefield-role sections (`BATTLELINE`, `CHARACTERS`,
 * `ALLIED UNITS`, …). Each unit is a header line `Name (N pts)` followed by
 * `•`-bulleted entries:
 *
 * ```
 * War Dog Executioner (130 pts)
 * • 1x Armoured feet
 * • 2x War Dog autocannon
 * • Houndpack Lance Character, Warlord
 *
 * Nurglings (40 pts)
 * • 3x Nurgling Swarm
 *     • 3x Diseased claws and teeth
 * ```
 *
 * Bullet classification (the parsing crux):
 * - A top-level `• Nx Thing` *with* further-indented child bullets is a **model
 *   group** — `N` adds to the model count and the children are that group's
 *   wargear (Nurglings, Beasts of Nurgle).
 * - A top-level `• Nx Thing` *without* children is plain **wargear**.
 * - A bullet *without* an `Nx` count is an **annotation**: `… Character` flags a
 *   character, `Warlord` flags the warlord, `Name (+N pts)` is the enhancement.
 *
 * **Disjointness from the WTC matchers**: the GW format always carries `•`
 * bullets and never the WTC `N with` lines. wtc-full always has `N with` (so it
 * never collides), and wtc-compact never has bullets (its matcher now excludes
 * them). This adapter therefore matches on *bullets present* + *no `N with`*.
 *
 * The GW export carries no separate POINTS LIMIT line, so `declared_limit`
 * falls back to TOTAL ARMY POINTS (the round-trippable battle-size signal).
 *
 * @packageDocumentation
 */
import type { FormatAdapter } from "./adapter.js";
import type { ParsedRoster, ParsedUnit, ParsedWargear } from "./types.js";
import {
  factionFromKeyword,
  inferBattleSizeRaw,
  stripParenthetical,
} from "./newrecruit-text.js";

const FACTION_KEYWORD_PREFIX = "+ FACTION KEYWORD:";

const HEADER_FIELDS = {
  faction: /^\+\s*FACTION KEYWORD:\s*(.+?)\s*$/i,
  detachment: /^\+\s*DETACHMENT:\s*(.+?)\s*$/i,
  totalPoints: /^\+\s*TOTAL ARMY POINTS:\s*(\d+)\s*pts?\s*$/i,
} as const;

const FENCE = /^\++\s*$/;
const HEADER_LINE = /^\+/;
const SECTION_HEADER = /^[A-Z][A-Z0-9 \-/&]+$/; // BATTLELINE, ALLIED UNITS, …
const UNIT_HEADER = /^(.+?)\s*\(\s*(\d+)\s*pts?\s*\)\s*$/i;
const BULLET_LINE = /^(\s*)•\s*(.+?)\s*$/u;
const NX_PREFIX = /^(\d+)x\s+(.+)$/;
const ENHANCEMENT_ANNOT = /^(.+?)\s*\(\+\s*(\d+)\s*pts?\s*\)\s*$/i;
const WITH_LINE = /^[\t ]*\d+\s+with\b/m;
const BULLET = /^[\t ]*•/mu;

const ALLIED_SECTION = "ALLIED UNITS";
const CHARACTERS_SECTION = "CHARACTERS";
const CHARACTER_SUFFIX = " Character";
const WARLORD_MARKER = "Warlord";

/** Accept the input only when it carries the FACTION KEYWORD summary header,
 * has `•` bullets, and lacks the WTC `N with` body lines. */
function isGwText(decoded: unknown): string | null {
  if (typeof decoded !== "string") return null;
  if (!decoded.includes(FACTION_KEYWORD_PREFIX)) return null;
  if (!BULLET.test(decoded)) return null;
  if (WITH_LINE.test(decoded)) return null; // that's wtc-full
  return decoded;
}

interface GwHeader {
  name: string;
  faction_raw_name: string | null;
  detachment_raw_name: string | null;
  total_reported: number | null;
  declared_limit: number | null;
  battle_size_raw: string | null;
}

function parseHeader(lines: string[]): { header: GwHeader; bodyStart: number } | null {
  let faction_raw_name: string | null = null;
  let detachment_raw_name: string | null = null;
  let total_reported: number | null = null;

  const fenceIndices: number[] = [];
  for (let i = 0; i < lines.length && fenceIndices.length < 2; i += 1) {
    if (FENCE.test(lines[i])) fenceIndices.push(i);
  }

  let sawFactionKeyword = false;
  for (const line of lines) {
    if (!line.startsWith("+")) continue;
    const factionMatch = HEADER_FIELDS.faction.exec(line);
    if (factionMatch) {
      faction_raw_name = factionFromKeyword(factionMatch[1]);
      sawFactionKeyword = true;
      continue;
    }
    const detMatch = HEADER_FIELDS.detachment.exec(line);
    if (detMatch) {
      detachment_raw_name = stripParenthetical(detMatch[1]);
      continue;
    }
    const ptsMatch = HEADER_FIELDS.totalPoints.exec(line);
    if (ptsMatch) {
      total_reported = Number.parseInt(ptsMatch[1], 10);
    }
  }

  if (!sawFactionKeyword) return null;

  const bodyStart = fenceIndices.length >= 2 ? fenceIndices[1] + 1 : 0;
  // The GW export has no POINTS LIMIT line — only TOTAL ARMY POINTS. Use it as
  // the declared limit so the inferred battle size stays round-trippable.
  const declared_limit = total_reported;
  return {
    header: {
      name: "Imported roster",
      faction_raw_name,
      detachment_raw_name,
      total_reported,
      declared_limit,
      battle_size_raw: inferBattleSizeRaw(declared_limit),
    },
    bodyStart,
  };
}

interface Bullet {
  indent: number;
  count: number | null;
  text: string;
}

interface UnitAcc {
  raw_name: string;
  displayed_pts: number | null;
  section: string | null;
  bullets: Bullet[];
}

function finishUnit(acc: UnitAcc): ParsedUnit {
  const topIndent = acc.bullets.length
    ? Math.min(...acc.bullets.map((b) => b.indent))
    : 0;

  const wargear = new Map<string, number>();
  let model_count = 0;
  let is_warlord = false;
  let is_character = acc.section === CHARACTERS_SECTION;
  let enhancement_raw_name: string | null = null;
  let enhancement_points: number | null = null;

  const addWargear = (raw_name: string, count: number): void => {
    wargear.set(raw_name, (wargear.get(raw_name) ?? 0) + count);
  };

  for (let i = 0; i < acc.bullets.length; i += 1) {
    const b = acc.bullets[i];

    // A child bullet (deeper than the unit's top level) is a model group's
    // weapon — its `Nx` count is already the squad-wide total.
    if (b.indent > topIndent) {
      if (b.count !== null) addWargear(b.text, b.count);
      continue;
    }

    // Top-level annotation (no `Nx` count): enhancement / character / warlord.
    if (b.count === null) {
      const enh = ENHANCEMENT_ANNOT.exec(b.text);
      if (enh) {
        if (enhancement_raw_name === null) {
          enhancement_raw_name = enh[1].trim();
          enhancement_points = Number.parseInt(enh[2], 10);
        }
        continue;
      }
      for (const token of b.text.split(",").map((s) => s.trim()).filter(Boolean)) {
        if (token === WARLORD_MARKER) is_warlord = true;
        else if (token.endsWith(CHARACTER_SUFFIX)) is_character = true;
      }
      continue;
    }

    // Top-level `Nx` bullet: a model group when it has child bullets beneath
    // it, otherwise plain wargear.
    const next = acc.bullets[i + 1];
    if (next && next.indent > topIndent) {
      model_count += b.count;
    } else {
      addWargear(b.text, b.count);
    }
  }

  if (model_count === 0) model_count = 1;

  // The GW unit header points include the enhancement; back it out to the base.
  const displayed = acc.displayed_pts;
  const points =
    displayed === null
      ? null
      : enhancement_points !== null
        ? displayed - enhancement_points
        : displayed;

  const wargearList: ParsedWargear[] = [];
  for (const [raw_name, count] of wargear) wargearList.push({ raw_name, count });

  return {
    raw_name: acc.raw_name,
    is_character,
    model_count,
    points,
    is_warlord,
    enhancement_raw_name,
    enhancement_points,
    wargear: wargearList,
  };
}

function parseBody(lines: string[], bodyStart: number): {
  units: ParsedUnit[];
  multi_force: boolean;
} {
  const units: ParsedUnit[] = [];
  let current: UnitAcc | null = null;
  let section: string | null = null;
  let alliedUnits = 0;

  const finalize = (): void => {
    if (current) {
      units.push(finishUnit(current));
      current = null;
    }
  };

  for (let i = bodyStart; i < lines.length; i += 1) {
    const raw = lines[i];
    const line = raw.trim();
    if (!line || FENCE.test(line) || HEADER_LINE.test(line)) continue;

    const bulletMatch = BULLET_LINE.exec(raw);
    if (bulletMatch) {
      if (current) {
        const indent = bulletMatch[1].length;
        const rest = bulletMatch[2];
        const nx = NX_PREFIX.exec(rest);
        current.bullets.push({
          indent,
          count: nx ? Number.parseInt(nx[1], 10) : null,
          text: (nx ? nx[2] : rest).trim(),
        });
      }
      continue;
    }

    const unitMatch = UNIT_HEADER.exec(line);
    if (unitMatch) {
      finalize();
      current = {
        raw_name: unitMatch[1].trim(),
        displayed_pts: Number.parseInt(unitMatch[2], 10),
        section,
        bullets: [],
      };
      if (section === ALLIED_SECTION) alliedUnits += 1;
      continue;
    }

    if (SECTION_HEADER.test(line)) {
      finalize();
      section = line;
    }
  }

  finalize();
  return { units, multi_force: alliedUnits > 0 };
}

export const gwAdapter: FormatAdapter = {
  id: "gw",

  matches(decoded: unknown): boolean {
    return isGwText(decoded) !== null;
  },

  parse(decoded: unknown): ParsedRoster {
    const text = isGwText(decoded);
    if (text === null) throw new Error("gw: input is not a GW app text export");

    const lines = text.split(/\r?\n/);
    const parsed = parseHeader(lines);
    if (!parsed) throw new Error('gw: missing "+ FACTION KEYWORD:" header');
    const { header, bodyStart } = parsed;

    const { units, multi_force } = parseBody(lines, bodyStart);

    let total_computed = 0;
    for (const u of units) {
      total_computed += u.points ?? 0;
      total_computed += u.enhancement_points ?? 0;
    }

    return {
      name: header.name,
      generated_by: null,
      faction_raw_name: header.faction_raw_name,
      detachment_raw_names: header.detachment_raw_name ? [header.detachment_raw_name] : [],
      battle_size_raw: header.battle_size_raw,
      declared_limit: header.declared_limit,
      total_reported: header.total_reported,
      total_computed,
      units,
      multi_force,
    };
  },
};

// Internals re-exported for unit tests.
export const _internals = {
  isGwText,
  parseHeader,
  parseBody,
};
